updating phone number lookups for the last 5 years

virtue3 opened the pull request:
I am slightly in disbelief that this is powering our phone number lookup.  But here we are.

For reference, the "real" fix for this should be to use twilios phonenumber lookup system.  It is free to determine the country/location of each number.  Mind you, we are then tied to twilio being up/down to do anything, but that's how it is with anything regarding SMS so I don't see the negative of this.

Tests incoming.

d-natoli commented:
Where did you get these updated numbers from?

virtue3 commented:
https://nationalnanpa.com/reports/area_code_relief_planning.html

It's showing at least one number as being US when it's Puerto Rico.

d-natoli commented on lib/e164/lookuptable.rb:
```diff
@@ -1,646 +1,690 @@
+# rubocop:disable Metrics/ModuleLength
+# rubocop:disable Naming/ConstantName
+# rubocop:disable Style/MutableConstant
+# rubocop:disable Style/WordArray
```
Don't disable any of these cops except for `ModuleLength`. You can run `rubocop -a` to autofix the `MutableConstant` and `WordArray` ones, and the `ConstantName` one will just require a replacement of `LookupTable` with `LOOKUP_TABLE`.

virtue3 replied:
I don't want to mess with this now because of comparing the previous tables.  

I tried the rubocop -a stuff and it actually broke a bunch of things on my first try.  I wasn't super happy with it

d-natoli commented on spec/e164_spec.rb:
```diff
@@ -3,11 +3,19 @@
+      expect(E164.lookup('14037238282')).to eq(['CA', 'Canada'])
     end
 
     it 'doesn\'t flail with a malformed phone number' do
```
`"doesn't fail with a malformed phone number"`

d-natoli commented on spec/e164_spec.rb:
```diff
@@ -3,11 +3,19 @@
-      expect(E164.lookup("@remindhq")).to eq(nil)
+      expect(E164.lookup('@remindhq')).to eq(nil)
+    end
+    it 'respects 689 as valid us country code' do
```
New line above this.

d-natoli commented on lib/e164/lookuptable.rb:
```diff
@@ -1,646 +1,690 @@
+    '1336' => ['US', 'United States'],
+    '1337' => ['US', 'United States'],
+    '1339' => ['US', 'United States'],
+    # '1340' => ['US', 'United States'], us virgin islands
```
All these commented ones had responses in the old format, e.g. this was `["VI", "Virgin Islands, US"]`. Should we just keep that the same?

virtue3 replied:
The old format one's are still there just down below.  This is entirely for keeping future diff's more sane.

d-natoli commented:
My concern is that we've removed some, and I'm not sure where we originally got these codes from. I'd like to get another opinion if this is fine to merge from @hamiltop.

hamiltop commented on lib/e164/lookuptable.rb:
```diff
@@ -1,10 +1,14 @@
+# this file is manually edited by pulling down the NAPA database and editing this:
+# https://nationalnanpa.com/reports/area_code_relief_planning.html
+
+# this only updated the US country codes.  Everything else is < 2015.  We generally only care if it's a USA phone number
```
Whoa... We absolutely care about canada codes.

virtue3 replied:
We don't in api.  The code is only checking for the country to match US.  So if we return nil or something else, it will fail the device validation.

```ruby
  def supported_country?(country)
    supported_countries.include?(country || 'US')
  end

  # Public: Returns true if the numeric country code is supported for SMS.
  #
  # code - The numeric (phone) country code.
  #
  # Returns Boolean.
  def supported_sms_region?(number)
    byebug

    # E164 is fundamentally flawd because it does a parse for country codes from the BACK of the string.
    # this means for a perfectly valid area code of 689 we get :
    # E164.lookup('6892109310')
    # ["PF", "French Polynesia"]
    # (byebug) E164.lookup('16892109310')
    # nil
    country_code, _ = E164.lookup(number.gsub(/\D/, '')) # E164 doesn't want formatting cruft
    country_code && (supported_country? country_code)
  end```

hamiltop commented on lib/e164/lookuptable.rb:
```diff
@@ -398,13 +348,88 @@ module E164
     "1985" => ["US", "United States"],
     "1986" => ["US", "United States"],
     "1989" => ["US", "United States"],
+    "1204" => ["CA", "Canada"],
```
Why are we no longer doing Canada inline?

hamiltop commented on lib/e164/lookuptable.rb:
```diff
@@ -398,13 +348,88 @@ module E164
+    "1888" => ["US", "Toll Free"],
+    "1902" => ["CA", "Canada"],
+    "1905" => ["CA", "Canada"],
+    "1939" => ["PR", "Puerto Rico"],
```
This one is now a duplicate entry.

virtue3 replied:
Old view of code. I specifically installed rubocop to lint duplicate hash values (there are none now).

virtue3 commented:
@hamiltop 

I agree that this isn't a prefect data set, and it's not intending to be.  I think the correct approach here is to use twilios lookup api because that is what ultimately matters.

I added in the source data here from NANP because that is the authoritative source.  I unfortunately, had to massage the data to remove the entries that are US territories or such that have different twilio pricing :/ . 

I went back through API and it is checking in one spot if it's US or CA before sending a marketing message.  Which seems horribly wrong but I'm going to go add those in now.  :/

virtue3 commented:
@d-natoli @hamiltop 

I've added some pretty strong tests that empirically prove the dataset is either the same, or purposefully changed.  (as well as tests to make sure I'm comparing the right data sets).  

This brought up a few changes (such as 1-900 and 1-911 that I have removed).  And a misclassified data item that is actually canada and not US.

virtue3 commented:
Also added new Canada numbers and re-ran tests (and passed).

hamiltop commented:
@tor can you post the effective data diff in a comment here?

virtue3 commented:
@hamiltop do you mean a full list of what's been added and/or removed?  And by here do you mean GitHub or the source?

hamiltop commented:
A full list of what's been added/changed/removed, and just here in a GitHub comment is fine.

virtue3 commented:
Removed 1911 and 1900 as valid US country area codes
1579 is set to canada properly.  

Here is the additions:
```
new area code 1227 with country info ["US", "United States"]
new area code 1274 with country info ["US", "United States"]
new area code 1283 with country info ["US", "United States"]
new area code 1326 with country info ["US", "United States"]
new area code 1327 with country info ["US", "United States"]
new area code 1341 with country info ["US", "United States"]
new area code 1353 with country info ["US", "United States"]
new area code 1369 with country info ["US", "United States"]
new area code 1447 with country info ["US", "United States"]
new area code 1448 with country info ["US", "United States"]
new area code 1464 with country info ["US", "United States"]
new area code 1557 with country info ["US", "United States"]
new area code 1564 with country info ["US", "United States"]
new area code 1582 with country info ["US", "United States"]
new area code 1627 with country info ["US", "United States"]
new area code 1659 with country info ["US", "United States"]
new area code 1679 with country info ["US", "United States"]
new area code 1689 with country info ["US", "United States"]
new area code 1730 with country info ["US", "United States"]
new area code 1764 with country info ["US", "United States"]
new area code 1835 with country info ["US", "United States"]
new area code 1839 with country info ["US", "United States"]
new area code 1840 with country info ["US", "United States"]
new area code 1975 with country info ["US", "United States"]
new area code 1257 with country info ["CA", "Canada"]
new area code 1263 with country info ["CA", "Canada"]
new area code 1273 with country info ["CA", "Canada"]
new area code 1354 with country info ["CA", "Canada"]
new area code 1368 with country info ["CA", "Canada"]
new area code 1382 with country info ["CA", "Canada"]
new area code 1387 with country info ["CA", "Canada"]
new area code 1428 with country info ["CA", "Canada"]
new area code 1460 with country info ["CA", "Canada"]
new area code 1468 with country info ["CA", "Canada"]
new area code 1474 with country info ["CA", "Canada"]
new area code 1487 with country info ["CA", "Canada"]
new area code 1537 with country info ["CA", "Canada"]
new area code 1568 with country info ["CA", "Canada"]
new area code 1584 with country info ["CA", "Canada"]
new area code 1683 with country info ["CA", "Canada"]
new area code 1742 with country info ["CA", "Canada"]
new area code 1753 with country info ["CA", "Canada"]
new area code 1851 with country info ["CA", "Canada"]
new area code 1871 with country info ["CA", "Canada"]
new area code 1879 with country info ["CA", "Canada"]
new area code 1942 with country info ["CA", "Canada"]
```
There was 46!!!! in total additions.
